x/pool-incentives/keeper/keeper.go

```diff
@@ -100,43 +100,73 @@ func (k Keeper) CreateConcentratedLiquidityPoolGauge(ctx sdk.Context, poolId uin
 		return fmt.Errorf("pool %d is not concentrated liquidity pool", poolId)
 	}
 
-	gaugeIdCL, err := k.incentivesKeeper.CreateGauge(
+	incentivesEpoch := k.incentivesKeeper.GetEpochInfo(ctx)
+
+	gaugeId, err := k.incentivesKeeper.CreateGauge(
 		ctx,
 		true,
 		k.accountKeeper.GetModuleAddress(types.ModuleName),
 		sdk.Coins{},
-		// dummy variable so that the existing logic does not break
-		// CreateGauge checks if LockQueryType is `ByDuration` or not, we bypass this check by passing
-		// lockQueryType as byTime. Although we do not need this check, we still cannot pass empty struct.
 		lockuptypes.QueryCondition{
-			LockQueryType: lockuptypes.ByTime,
-			Denom:         appParams.BaseCoinUnit,
+			LockQueryType: lockuptypes.NoLock,
+			Denom:         incentivestypes.NoLockInternalGaugeDenom(pool.GetId()),
+			Duration:      incentivesEpoch.Duration,
 		},
 		ctx.BlockTime(),
 		1,
+		poolId,
 	)
 	if err != nil {
 		return err
 	}
 
-	incParams := k.incentivesKeeper.GetEpochInfo(ctx)
-	// lockable duration is epoch duration because we create incentive_record on every epoch
-	k.SetPoolGaugeId(ctx, poolId, incParams.Duration, gaugeIdCL)
+	// Although the pool id <> gauge "NoLock" link is created in CreateGauge,
+	// we create an additional "ByDuration" link here for tracking
+	// internal incentive "NoLock" gauges
+	incentivesEpochDuration := k.incentivesKeeper.GetEpochInfo(ctx).Duration
+	k.SetPoolGaugeIdInternalIncentive(ctx, poolId, incentivesEpochDuration, gaugeId)
 
 	return nil
 }
 
-func (k Keeper) SetPoolGaugeId(ctx sdk.Context, poolId uint64, lockableDuration time.Duration, gaugeId uint64) {
-	key := types.GetPoolGaugeIdStoreKey(poolId, lockableDuration)
+// SetPoolGaugeIdInternalIncentive sets the gauge id for the pool and internally incentivized duration.
+// CONTRACT: this link is created only for the internally incentivized gauges.
+func (k Keeper) SetPoolGaugeIdInternalIncentive(ctx sdk.Context, poolId uint64, incentivizedDuration time.Duration, gaugeId uint64) {
+	// Note: this index is used for internal incentive gauges only.
+	key := types.GetPoolGaugeIdInternalStoreKey(poolId, incentivizedDuration)
+	store := ctx.KVStore(k.storeKey)
+	store.Set(key, sdk.Uint64ToBigEndian(gaugeId))
+
+	// Note: this index is used for general linking.
+	key = types.GetPoolIdFromGaugeIdStoreKey(gaugeId, incentivizedDuration)
+	store.Set(key, sdk.Uint64ToBigEndian(poolId))
+}
+
+// SetPoolGaugeIdNoLock sets the link between pool id and gauge id for "NoLock" gauges.
+// CONTRACT: the gauge of the given id must be "NoLock" gauge.
+func (k Keeper) SetPoolGaugeIdNoLock(ctx sdk.Context, poolId uint64, gaugeId uint64) {
```

Comment:
What specifically are you suggesting? I currently don't see anything that can be done where the added complexity would be justified by the security benefit but open to try specific paths if you see any

Reply:
I was thinking of a case where with these locable Duration [0s, 1m, 1h] in that case there will be 2 store entries with same Key no? 

here: https://github.com/osmosis-labs/osmosis/pull/5459/files#diff-968d496699ec6c793045c4d26a3a3de4af2f46e5a5d9a94b81ae11e0e59a1903R142

and here: https://github.com/osmosis-labs/osmosis/pull/5459/files#diff-968d496699ec6c793045c4d26a3a3de4af2f46e5a5d9a94b81ae11e0e59a1903R161

Reply:
It is idempotent. If one is written, the other one will have no effect

Reply:
I see, not really concerned about that case because we explicitly hardcode that value to be incentivized epoch duration for "No Lock" gauges, changing which is gated by governance.

However, I will add the suggested check prior to merge